### PR TITLE
Make relative paths work in various places

### LINKED ISSUE #71: obj_dir command line arg support relative paths
It seems like the only command line arg that doesn't, -p and -e both do.

### PR DESCRIPTION
**Pull request by jwright6323**
**Metadata: rebase merge**

This will aim to fix #71, #270, ~and #276~. I'm going to try to fix them individually so let's rebase merge this.

### CODE REVIEW

**Comment by edwardcwang** [comment]
It seems like #270 and #276 are separate from this PR though...

**Comment by jwright6323** [comment]
I haven't gotten there yet.

**Comment by edwardcwang** [comment]
OK. Would appreciate it if we could follow the plan sketched out in those issues rather than reinventing something new.

**Comment by jwright6323** [comment]
That is the plan

**Comment by jwright6323** [comment]
I'm not sure I understand the `abspath` stuff so I'm going to ignore #276 for now.

**Comment by jwright6323** [comment]
This is ready for review (not fixing #276). See #458 for something I ran into while writing the tests.

**Comment by jwright6323** [comment]
Somehow I've introduced a python3.3-3.5 incompatibility :(

**Comment by jwright6323** [comment]
...I have no idea what I did to fix this. I just removed accidental `print`s and it started working.

**Comment by jwright6323** [comment]
The problem was a real bug that only showed up depending on the iteration order a python dict. I guess that behavior changed from 3.5 to 3.6 masking the bug.

**Comment by edwardcwang** [comment]
IIRC dictionary orderings are not to be trusted anyway...

**Comment by jwright6323** [comment]
> IIRC dictionary orderings are not to be trusted anyway...

Yep, it was actually that a particular ordering masked a bug (the wrong dictionary was used, and in some cases it had the requisite keys, depending on the iteration order).

File: src/hammer_config_test/test.py [hunk 1]
```diff
@@ -801,6 +801,106 @@ def test_self_reference_lazyappend(self) -> None:
...
+        meta = hammer_config.load_config_from_string("""
+{
+  "foo.bar_meta": ["append", "deepsubst"],
+  "foo.bar": [{
+      "dsl": ["python"],
+      "dsl_meta": "append",
+      "dac": "current_weighted",
+      "meta_test": "local_path",
+      "meta_test_deepsubst_meta": "local",
+      "abs_test": "/this/is/an/abs/path",
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config_test/test.py]: `os.path.join` does it automatically (as I came to find out):

```
>>> os.path.join("foo", "/this/is/an/abspath")
'/this/is/an/abspath'
```

File: src/hammer_config/config_src.py [hunk 2]
```diff
@@ -354,6 +380,88 @@ def prependlocal_rename(key: str, value: Any, target_setting: str, replacement_s
...
+                                    if isinstance(v, str):
+                                        newval[k] = DeepSubstMetaDirectives[meta](v, params)
+                                    else:
+                                        raise ValueError("Deepsubst metas not supported on non-string values: {}".format(str(v)))
+                                else:
+                                    raise ValueError("Unknown deepsubst_meta type: {}".format(str(v)))
+                            else:
+                                newval[k] = do_subst(v)
+                    else:
+                        # Will this ever happen???
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: Sure, I can add some more comments.

File: src/hammer_config/config_src.py [hunk 3]
```diff
@@ -354,6 +380,88 @@ def prependlocal_rename(key: str, value: Any, target_setting: str, replacement_s
...
+                # This is just regular subst
+                return subst_str(oldval, lambda key: config_dict[key])
+            elif isinstance(oldval, list):
+                return list(map(do_subst, oldval))
+            elif isinstance(oldval, dict):
+                # We need to check for _deepsubst_meta here
+                newval = {}  # type: Dict
+                for k,v in oldval.items():
+                    if isinstance(k, str):
+                        m = re.match(r"^(.*)_deepsubst_meta$", k)
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: I can change it if you want, but then I have to strip the suffix off some other way- this does both.

File: src/hammer_config/config_src.py [hunk 4]
```diff
@@ -354,6 +380,88 @@ def prependlocal_rename(key: str, value: Any, target_setting: str, replacement_s
...
+        """
+        Perform a deep substitution on the value provided. This will replace any variables that occur in strings
+        of the form ${...} and will also do a special meta replacement on keys which end in _deepsubst_meta.
+        """
+        def do_subst(oldval: Any) -> Any:
+            if isinstance(oldval, str):
+                # This is just regular subst
+                return subst_str(oldval, lambda key: config_dict[key])
+            elif isinstance(oldval, list):
+                return list(map(do_subst, oldval))
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: Just for you :)

File: src/hammer_config/config_src.py [hunk 5]
```diff
@@ -91,6 +91,32 @@ class MetaDirective(NamedTuple('MetaDirective', [
...
+def deepsubst_cwd(path: str, params: MetaDirectiveParams) -> str:
+    """
+    Prepend the current working directory (of the hammer runtime) to the beginning of the
+    specified path.
+
+    :param path: The string path to which the CWD is to be prepended.
+    :param params: The MetaDirectiveParams (not used by this method).
+    :return: The path with CWD prepended.
+    """
+    return os.path.join(os.getcwd(), path)
```
Comment (edwardcwang) [inline src/hammer_config/config_src.py]:
```suggestion
    # os.path.join handles the case where path is absolute
    # "If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component."
    return os.path.join(os.getcwd(), path)
```

File: src/hammer_config/config_src.py [hunk 6]
```diff
@@ -91,6 +91,32 @@ class MetaDirective(NamedTuple('MetaDirective', [
...
+def deepsubst_local(path: str, params: MetaDirectiveParams) -> str:
+    """
+    Prepend the directory containing the config file containing this setting to the
+    beginning of the specified path.
+
+    :param path: The string path to which the local path is to be prepended.
+    :param params: The MetaDirectiveParams which contain the local path.
+    :return: The path with the local path of the config prepended.
+    """
+    return os.path.join(params.meta_path, path)
```
Comment (edwardcwang) [inline src/hammer_config/config_src.py]:
```suggestion
    # os.path.join handles the case where path is absolute
    # "If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component."
    return os.path.join(params.meta_path, path)
```

File: src/hammer_config/config_src.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: Done

File: src/hammer_config/config_src.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (edwardcwang) [reply src/hammer_config/config_src.py]: Hmm. My preference would be to use simpler string methods as we aren't using the full complexity of a regex...

[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: This is a pretty simple regex. Can you suggest how to remove the suffix?

[the remark this answers is not in the document]
Comment (colinschmidt) [reply src/hammer_config/config_src.py]: k.repalce("_deepsubst_meta","")`

[the remark this answers is not in the document]
Comment (jwright6323) [reply src/hammer_config/config_src.py]: OK, fixed

**Comment by colinschmidt** [comment]
So happy to finally have this resolved! Thanks John!

**Pull request merged**